File: examples/LMemLoopback/client/java/AdvancedStatic/LMemLoopbackClient.java
import com.maxeler.LMemLoopback.LMemLoopbackService;
import com.maxeler.LMemLoopback.LMemLoopback_actions_t_struct;
import com.maxeler.LMemLoopback.LMemLoopback_readLMem_actions_t_struct;
import com.maxeler.LMemLoopback.LMemLoopback_writeLMem_actions_t_struct;

import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

/** LMemLoopback AdvancedStatic example. */
public final class LMemLoopbackClient {

  /** Server port. */
  private static final int PORT = 9090;

  /** Size of int in bytes. */
  private static final int SIZE_OF_INT_IN_BYTES = 4;

  /** Size of int in bits. */
  private static final int SIZE_OF_INT = 32;

  /** Number of nano seconds in one second. */
  private static final int NUM_OF_NANO_SECONDS = 1000000000;

  /** Utility classes should not have a public or default constructor. */
  private LMemLoopbackClient() { }

  /**
   * Checks if lmemLoopbackDfe and lmemLoopbackCpu are the same.
   *
   * @param size       Size
   * @param dataOutDfe Out data from DFE
   * @param dataOutCpu Out data from CPU
   *
   * @return status Number of elements that doesn't match
   */
  public static int check(final List<Integer> dataOutDfe,
                           final List<Integer> dataOutCpu, final int size) {
    int status = 0;

    for (int i = 0; i < size; i++) {
      if (!dataOutDfe.get(i).equals(dataOutCpu.get(i))) {
        System.out.println(
            "Output data @ " + i + " = " + dataOutDfe.get(i)
            + " (expected " + dataOutCpu.get(i) + ")");

        status++;
      }
    }

    return status;
  }

  /**
   * LMemLoopback on CPU. outData = inA + inB
   *
   * @param size    Size
   * @param inA     In A
   * @param inB     In B
   *
   * @return        Out data
   */
  public static List<Integer> lmemLoopbackCpu(
      final int size, final List<Integer> inA, final List<Integer> inB) {
    List<Integer> outData = new ArrayList<Integer>();

    for (int i = 0; i < size; i++) {
      outData.add(inA.get(i) + inB.get(i));
    }

    return outData;
  }

  /**
   * LMemLoopback on DFE. outData = inA + inB
   *
   * @param size    Size
   * @param inA     In A
   * @param inB     In B
   *
   * @return        Out data
   */
  public static List<Integer> lmemLoopbackDfe(
      final int size, final List<Integer> inA, final List<Integer> inB) {
    List<Integer> outData = new ArrayList<Integer>();
    DecimalFormat timeFormat = new DecimalFormat("#0.00000");

    double startTime = System.nanoTime();
    double estimatedTime;
    int sizeBytes = size * SIZE_OF_INT_IN_BYTES;

    // Make socket
    TTransport transport;
    transport = new TSocket("localhost", PORT);

    // Wrap in a protocol
    TProtocol protocol = new  TBinaryProtocol(transport);

    // Create a client to use the protocol encoder
    LMemLoopbackService.Client client =
        new LMemLoopbackService.Client(protocol);

    estimatedTime = (System.nanoTime() - startTime)
                         / NUM_OF_NANO_SECONDS;
    System.out.println("Creating a client:\t\t\t\t\t"
                       + timeFormat.format(estimatedTime) + "s");
    try {
      // Connect!
      startTime = System.nanoTime();
      transport.open();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Opening connection:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Initialize maxfile
      startTime = System.nanoTime();
      final long maxfile = client.LMemLoopback_init();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Initializing maxfile:\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Load DFE
      startTime = System.nanoTime();
      final long engine = client.max_load(maxfile, "*");
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Loading DFE:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Allocate and send input streams to server
      startTime = System.nanoTime();
      final long addressInA = client.malloc_int32_t(size);
      client.send_data_int32_t(addressInA, inA);

      final long addressInB = client.malloc_int32_t(size);
      client.send_data_int32_t(addressInB, inB);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Sending input data:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Allocate memory for output stream on server
      startTime = System.nanoTime();
      final long addressOutData = client.malloc_int32_t(size);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Allocating memory for output stream on server:\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Write to LMem
      startTime = System.nanoTime();
      LMemLoopback_writeLMem_actions_t_struct actionsLmemWrite =
          new LMemLoopback_writeLMem_actions_t_struct();

      actionsLmemWrite.setParam_address(0);
      actionsLmemWrite.setParam_nbytes(sizeBytes);
      actionsLmemWrite.setInstream_cpu_to_lmem(addressInA);

      long addressActionsLmemWrite =
          client.send_LMemLoopback_writeLMem_actions_t(actionsLmemWrite);
      client.LMemLoopback_writeLMem_run(engine, addressActionsLmemWrite);

      actionsLmemWrite.setParam_address(sizeBytes);
      actionsLmemWrite.setParam_nbytes(sizeBytes);
      actionsLmemWrite.setInstream_cpu_to_lmem(addressInB);

      addressActionsLmemWrite =
          client.send_LMemLoopback_writeLMem_actions_t(actionsLmemWrite);
      client.LMemLoopback_writeLMem_run(engine, addressActionsLmemWrite);

      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Writing to LMem:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Executing action
      startTime = System.nanoTime();
      LMemLoopback_actions_t_struct actions =
          new LMemLoopback_actions_t_struct();

      actions.setParam_N(size);

      final long addressActions = client.send_LMemLoopback_actions_t(actions);
      client.LMemLoopback_run(engine, addressActions);

      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("LMemLoopback time:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Read from LMem
      startTime = System.nanoTime();
      LMemLoopback_readLMem_actions_t_struct actionsLmemRead =
          new LMemLoopback_readLMem_actions_t_struct();

      actionsLmemRead.setParam_address(2 * sizeBytes);
      actionsLmemRead.setParam_nbytes(sizeBytes);
      actionsLmemRead.setOutstream_lmem_to_cpu(addressOutData);

      final long addressactionsLmemRead =
          client.send_LMemLoopback_readLMem_actions_t(actionsLmemRead);
      client.LMemLoopback_readLMem_run(engine, addressactionsLmemRead);

      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Reading from LMem:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Unload DFE
      startTime = System.nanoTime();
      client.max_unload(engine);

      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Unloading DFE:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Get output stream from server
      startTime = System.nanoTime();
      outData = client.receive_data_int32_t(addressOutData, size);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Getting output stream:\t(size = "
                          + size * SIZE_OF_INT + " bit)\t"
                          + timeFormat.format(estimatedTime) + "s");

      // Free allocated memory for streams on server
      startTime = System.nanoTime();
      client.free(addressInA);
      client.free(addressInB);
      client.free(addressOutData);
      client.free(addressActionsLmemWrite);
      client.free(addressActions);
      client.free(addressactionsLmemRead);
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated memory for streams on server:\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Free allocated maxfile data
      startTime = System.nanoTime();
      client.LMemLoopback_free();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Freeing allocated maxfile data:\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

      // Close!
      startTime = System.nanoTime();
      transport.close();
      estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
      System.out.println("Closing connection:\t\t\t\t\t"
                           + timeFormat.format(estimatedTime) + "s");

    } catch (TException x) {
      x.printStackTrace();
      System.exit(-1);
    }

    return outData;
  }

  /**
   * Calculates lmemLoopbackDfe and lmemLoopbackCpu and
   * checks if they return the same value.
   *
   * @param args Command line arguments
   */
  public static void main(final String[] args) {

    final int size = 384;
    double startTime = System.nanoTime();
    double estimatedTime;
    final DecimalFormat timeFormat = new DecimalFormat("#0.00000");

    // Generate data
    startTime = System.nanoTime();
    List<Integer> inA = new ArrayList<Integer>();
    List<Integer> inB = new ArrayList<Integer>();

    for (int i = 0; i < size; i++) {
      inA.add(i);
      inB.add(size - i);
    }
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("Generating input data:\t\t\t\t"
                         + timeFormat.format(estimatedTime)
                         + "s");
    // DFE Output
    startTime = System.nanoTime();
    final List<Integer> dataOutDfe = lmemLoopbackDfe(size, inA, inB);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("DFE LMemLoopback total time:\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    // CPU Output
    startTime = System.nanoTime();
    final List<Integer> dataOutCpu = lmemLoopbackCpu(size, inA, inB);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("CPU LMemLoopback total time:\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");

    // Checking results
    startTime = System.nanoTime();
    int status = check(dataOutDfe, dataOutCpu, size);
    estimatedTime = (System.nanoTime() - startTime) / NUM_OF_NANO_SECONDS;
    System.out.println("Checking results:\t\t\t\t\t"
                         + timeFormat.format(estimatedTime) + "s");
    if (status == 0) {
      System.out.println("Test passed!");
    } else {
      System.out.println("Test failed " + status + " times!");
      System.exit(-1);
    }
  }
}

File: examples/LMemLoopback/client/erl/BasicStatic/lMemLoopbackClient.erl
-module(lMemLoopbackClient).

-include("lMemLoopbackService_thrift.hrl").

-export([t/0]).

check ([], [], Status, _)                                                                                    ->    Status;
check (OutDFE, OutCPU, Status, Iter) when (hd(OutDFE) - hd(OutCPU)) * (hd(OutDFE) - hd(OutCPU)) > 0.00001    ->    io:format("Output data @ ~w = ~w (expected ~w)~n", [Iter, hd(OutDFE), hd(OutCPU)]),
                                                                                                                   check (tl(OutDFE), tl(OutCPU), (Status + 1), (Iter + 1));
check (OutDFE, OutCPU, Status, Iter) when (hd(OutDFE) - hd(OutCPU)) * (hd(OutDFE) - hd(OutCPU)) < 0.00001    ->    check (tl(OutDFE), tl(OutCPU), Status, Iter + 1) .

lMemLoopBackCPU ([], [])     ->    [] ;
lMemLoopBackCPU ([A | InA], [B | InB])    ->    [A + B | lMemLoopBackCPU (InA, InB)] .

lMemLoopBackDFE(Size, InA, InB) ->
    DataOutDFE = 
        try
            SizeBytes = Size * 4,
            StartTime0 = erlang:timestamp(),
            Port = 9090,
            {ok, Client0} = thrift_client_util:new("localhost",
                                                   Port,
                                                   lMemLoopbackService_thrift,
                                                   []),
            io:format("Creating a client and opening connection:\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime0) / 1000000]),

            % Allocate and send input streams to server
            StartTime1 = erlang:timestamp(),    
            {Client1, {ok, Address_inA}} = thrift_client:call(Client0, malloc_int32_t, [Size]),
            {Client2, {ok, ok}} = thrift_client:call(Client1, send_data_int32_t, [Address_inA, InA]),

            {Client3, {ok, Address_inB}} = thrift_client:call(Client2, malloc_int32_t, [Size]),
            {Client4, {ok, ok}} = thrift_client:call(Client3, send_data_int32_t, [Address_inB, InB]),
            io:format("Sending input data:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime1) / 1000000]),

            % Allocate memory for output stream on server
            StartTime2 = erlang:timestamp(),
            {Client5, {ok, Address_dataOut}} = thrift_client:call(Client4, malloc_int32_t, [Size]),
            io:format("Allocating memory for output stream on server:\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime2) / 1000000]),

            % Writing to LMem
            StartTime3 = erlang:timestamp(),
            {Client6, {ok, ok}} = thrift_client:call(Client5, 'LMemLoopback_writeLMem' , [0, SizeBytes, Address_inA]),
            {Client7, {ok, ok}} = thrift_client:call(Client6, 'LMemLoopback_writeLMem' , [SizeBytes, SizeBytes, Address_inB]),
            io:format("Writing to LMmem:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime3) / 1000000]),

            % Action default
            StartTime4 = erlang:timestamp(),
            {Client8, {ok, ok}} = thrift_client:call(Client7, 'LMemLoopback', [Size]),
            io:format("LMemLoopback time:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime4) / 1000000]),

            % Reading from LMem
            StartTime5 = erlang:timestamp(),
            {Client9, {ok, ok}} = thrift_client:call(Client8, 'LMemLoopback_readLMem' , [2 * SizeBytes, SizeBytes, Address_dataOut]),
            io:format("Reading from  LMmem:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime5) / 1000000]),

            % Get output stream from server
            StartTime6 = erlang:timestamp(),
            {Client10, {ok, DataOut}} = thrift_client:call(Client9, receive_data_int32_t, [Address_dataOut, Size]),
            io:format("Getting output stream:\t(size = ~ws bit)\t~ws~n", [Size * 32, timer:now_diff(erlang:timestamp(), StartTime6) / 1000000]),

            % Free allocated memory for streams on server
            StartTime7 = erlang:timestamp(),
            {Client11, {ok, ok}} = thrift_client:call(Client10, free, [Address_inA]),
            {Client12, {ok, ok}} = thrift_client:call(Client11, free, [Address_inB]),
            {Client13, {ok, ok}} = thrift_client:call(Client12, free, [Address_dataOut]),
            io:format("Freeing allocated memory for streams on server:\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime7) / 1000000]),

            % Close!
            StartTime8 = erlang:timestamp(),
            {_Client14, ok} = thrift_client:close(Client13),
            io:format("Closing connection:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime8) / 1000000]),

            DataOut
        catch
            error:Reason ->
                io:format("Error reason: ~p~n", [Reason]),
                erlang:exit(-1)
        end,
    
        DataOutDFE.

t() ->
    StartTime1 = erlang:timestamp(),
    Size = 384,
    InA = lists:seq(0, (Size-1)),
    InB = [ Size - A || A <- InA],
    io:format("Generating input data:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime1) / 1000000]),

    % DFE Output
    StartTime2 = erlang:timestamp(),
    DataOutDFE = lMemLoopBackDFE(Size, InA, InB),
    io:format("DFE LMemLoopback total time:\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime2) / 1000000]),

    % CPU Output
    StartTime3 = erlang:timestamp(),
    DataOutCPU = lMemLoopBackCPU(InA, InB),
    io:format("CPU LMemLoopback total time:\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime3) / 1000000]),

    % Checking results
    StartTime4 = erlang:timestamp(),
    Status = check(DataOutDFE, DataOutCPU, 0, 0),
    io:format("Checking results:\t\t\t\t~ws~n", [timer:now_diff(erlang:timestamp(), StartTime4) / 1000000]),

    if Status =:= 0    ->    io:format("Test successful!~n");
       true            ->    io:format("Test failed ~w times!~n", [Status]),
                             erlang:exit(-1)
    end,
    
    ok.
